### PR TITLE
[RFC]refactor: refact the container lock

### PR DESCRIPTION
**Pull request by HusterWan**
Signed-off-by: Michael Wan <zirenwan@gmail.com>

### Ⅰ. Describe what this PR did

Since the single container APIs have not occurred performance problem and the Fine-grained locking mechanism has occurred many concurrency problems, so we decide to change to Coarse-grained locking mechanism that will lock all operations in one API function. 

Of course, we should make sure this change has no influence on the legacy containers and the CRI interfaces, so we will first make all test cases happy, then consider to merge this pr.

### CODE REVIEW

File: daemon/mgr/container.go [hunk 1]
```diff
@@ -607,9 +602,18 @@ func (mgr *ContainerManager) Start(ctx context.Context, id string, options *type
...
+	// NOTE: add a big lock when start a container
+	c.Lock()
+	defer c.Unlock()
+
 	var err error
 	c.DetachKeys = options.DetachKeys
 
+	// check if container's status is running
+	if c.State.Running {
+		return errors.Wrapf(errtypes.ErrNotModified, "container already started")
```
Comment (Ace-Tang) [inline daemon/mgr/container.go]: I check docker start a running state container should not throw error, why we here

Comment (HusterWan) [reply daemon/mgr/container.go]: Actually no specific reasons, we can make the result staying with the moby

Comment (Ace-Tang) [reply daemon/mgr/container.go]: Sure

**Comment by Ace-Tang** [comment]
Maybe get deadlock somewhere or other reason, ci hang long make test fail

File: daemon/mgr/container.go (hunk 1 above)
Comment (HusterWan) [reply daemon/mgr/container.go]: > I check docker start a running state container should not throw an error, why we here

which docker version do you check? I check the lastest moby and 1.12.6 docker, they all throw an error when starting an already running container.

**Comment by HusterWan** [comment]
> Maybe get deadlock somewhere or other reason, ci hang long make test fail

@Ace-Tang yes, I will figure it out

File: daemon/mgr/container.go (hunk 1 above)
Comment (Ace-Tang) [reply daemon/mgr/container.go]: docker 1.12.6 version
```
[root@r10e19288.sqa.zmf /root]
#docker start caff7e06c83d
caff7e06c83d

[root@r10e19288.sqa.zmf /root]
#docker start caff7e06c83d
caff7e06c83d
```

Comment (Ace-Tang) [reply daemon/mgr/container.go]: Oh, I know, that error not throw in cli 
```
time="2019-03-18T09:52:13.025216482+08:00" level=info msg="Calling POST /v1.24/containers/caff7e06c83d/start, client @"
time="2019-03-18T09:52:13.025397258+08:00" level=error msg="Handler for POST /v1.24/containers/caff7e06c83d/start, client @ returned error: Container already started"
```

**Comment by Ace-Tang** [comment]
@HusterWan , does ths ready for review now?

**Comment by HusterWan** [comment]
> @HusterWan , does ths ready for review now?

Yes, you can review now.

**Comment by wangforthinker** [comment]
@HusterWan As we know, read ops are more than write ops, so rwlock is a good suggestion.

**Comment by HusterWan** [comment]
> @HusterWan As we know, read ops are more than write ops, so rwlock is a good suggestion.

@wangforthinker agree with you. actually, we only have one container read ops, just `Get` api. and we do not add lock for this api.

File: apis/types/update_config.go [hunk 2]
```diff
@@ -19,7 +19,9 @@ type UpdateConfig struct {
 	// update disk quota for container
 	DiskQuota map[string]string `json:"DiskQuota,omitempty"`
 
-	// A list of environment variables to set inside the container in the form `["VAR=value", ...]`. A variable without `=` is removed from the environment, rather than to have an empty value.
+	// A list of environment variables to set inside the container in the form `["VAR=value", ...]`.
```
Comment (rudyfly) [inline apis/types/update_config.go]: 这个改动好像rebase一下就可以了

**Comment by fuweid** [comment]
I think we should add lock for Create. Since we set the container name at the ending of function, it might create two containers with the same name. WDYT?

File: apis/types/update_config.go (hunk 2 above)
Comment (HusterWan) [reply apis/types/update_config.go]: good catch!, rebased

**Comment by HusterWan** [comment]
> I think we should add lock for Create. Since we set the container name at the ending of function, it might create two containers with the same name. WDYT?

I think we should change the time when to put name->id to the cache in case we may create containers with the same name. Using container Lock cannot fix the problem.

**Comment by HusterWan** [comment]
> > I think we should add lock for Create. Since we set the container name at the ending of function, it might create two containers with the same name. WDYT?
> 
> I think we should change the time when to put name->id to the cache in case we may create containers with the same name. Using container Lock cannot fix the problem.

I can create a new pr to fix this problem.

**Comment by Ace-Tang** [comment]
@HusterWan ， please rebase, since I fix a bug in https://github.com/alibaba/pouch/pull/2757
and  parallel test is pass.

**Comment by HusterWan** [comment]
> @HusterWan ， please rebase, since I fix a bug in #2757
> and parallel test is pass.

done

**Comment by allencloud** [comment]
Please solve the conflict and sync the latest code, and then we can move to consider merging. @HusterWan

**Pull request merged**